### ISSUE TITLE
SKIP: Should we support 32-bits architectures?

### ISSUE THREAD

**Issue by sciunto** [scikit-image/scikit-image#4430]
## Description

During the last dev meeting (today), we discussed with @emmanuelle and @lagru the support of 32-bits architecture.

The starting point is the number of issues involving this aspect, which are not solved mainly because most (all) of us are not using 32-bits. To get an idea: https://github.com/scikit-image/scikit-image/labels/type%3A%20failure%20arch%20specific

As pointed out by @emmanuelle if we do not have a job in our CI with this architecture, we can consider it as unsupported because we will never know whether it passes or fails.

@lagru found that azure is supporting 32-bits, so it would be possible to support it.

Now the question is: is this worth the effort? @scikit-image/core The debate is open :)

**Comment by jni** [comment]
What are NumPy and SciPy doing here? We should certainly go no further than they do. I've heard chatter of various projects dropping 32-bit. Conda-forge I think?

Personally, I think this issue needs a "champion" who cares about this (Bit Keeper?), and I'm not inclined to become that.

**Comment by hmaarrfk** [comment]
The thing is that matplotlib dropped 32 bit wheels for Linux. As a hard dependency, this makes building the wheels tedious since we need to pull in things from a wheel house.

I think windows 32 bit is still popular. Many apps still download 32 bit windows by default especially for older computers.

I would be pretty happy to drop it though.

**Comment by alexdesiqueira** [comment]
Hey everyone,

> What are NumPy and SciPy doing here? We should certainly go no
> further than they do. I've heard chatter of various projects
> dropping 32-bit. Conda-forge I think?

NumPy and SciPy still provide 32 bit wheels (..._i686.whl for Linux,
..._win32.whl for Windows):
[1] https://pypi.org/project/numpy/#files
[2] https://pypi.org/project/scipy/#files

> The thing is that matplotlib dropped 32 bit wheels for Linux. As a
> hard dependency, this makes building the wheels tedious since we
> need to pull in things from a wheel house.

Mark got a point. No 32-bit matplotlib for Linux anymore; however, they
still have Windows's.
[3] https://pypi.org/project/matplotlib/#files

Couldn't we do the same? Maybe maintain a 32-bit wheel only for
Windows, during a cycle or two, and start to warn the users we are
dropping support for 32-bit wheels on 20XX?
Kind regards,

Alex


On Fri, 2020-01-31 at 22:14 -0800, Mark Harfouche wrote:
> The thing is that matplotlib dropped 32 bit wheels for Linux. As a
> hard dependency, this makes building the wheels tedious since we need
> to pull in things from a wheel house.
> 
> I think windows 32 bit is still popular. Many apps still download 32
> bit windows by default especially for older computers.
> 
> I would be pretty happy to drop it though.
> 
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
-- 
--------------------------------------------------
Dr. Alexandre de Siqueira
Berkeley Institute for Data Science - BIDS
190 Doe Library
University of California, Berkeley
Berkeley, CA 94720
United States

Lattes CV: 3936721630855880
ORCID: 0000-0003-1320-4347
Github: alexdesiqueira
Twitter: alexdesiqueira
--------------------------------------------------

**Comment by hmaarrfk** [comment]
While 32 bit x86 has been dropped for Linux by many large distros, Rasbian still ships 32bit ARM by default.

I've seen Raspbian become an important platform for many getting into machine vision and would hate to see scikit-image being booted from that platform.

I don't think it is impossible to add a 32 bit platform to our test matrix. I would really appreciate help from matplotlib in making 32 bit wheels available on PyPi for Linux, but I did have a PR at some point in time that removed the matplotlib dependency so that is an other option.

**Comment by alexdesiqueira** [comment]
> I would be pretty happy to drop it though.
> 
> (...)
> 
> I've seen Raspbian become an important platform for many getting into machine vision and would hate to see scikit-image being booted from that platform.
> 
> I don't think it is impossible to add a 32 bit platform to our test matrix.

Please help me here Mark :sweat_smile: 

Starting 2016, Raspberry Pi have [both 32 and 64 bit versions](https://en.wikipedia.org/wiki/Raspberry_Pi#Specifications). I think this will be in the past for embedded stuff soon, as well.

**Comment by hmaarrfk** [comment]
It is true that the hardware supports 64 bit, however, the default os is still 32 bit

https://raspberrypi.stackexchange.com/questions/43921/will-raspbian-move-to-64-bit#43927

https://www.piwheels.org/project/scikit-image/

https://unix.stackexchange.com/a/136519
Note that armv7 is 32 bit

**Comment by matthew-brett** [comment]
I think it will be a while before Windows 32-bit goes away - see https://github.com/pypa/manylinux/issues/118#issuecomment-481404761

I think it's been true in the past that 32-bit errors and crashes have revealed coding errors that needed fixing for other platforms.

**Comment by matthew-brett** [comment]
I think the default Windows download from Python.org is still 32-bit.

**Comment by alexdesiqueira** [comment]
> It is true that the hardware supports 64 bit, however, the default os is still 32 bit
> (...)
> Note that armv7 is 32 bit
>
> I think the default Windows download from Python.org is still 32-bit.

Yes, you're both right. I guess it will still stick for a while.
It seems the optimal solution would be to remove matplotlib as a hard dep and continue to provide 32-bit wheels/tests. Thoughts?

**Comment by tacaswell** [comment]
I do not think it is fair to say "Matplotlib does not support 32bit", we just have not uploaded 32bit x86 wheels for linux.  To my knowledge, this is the first time anyone has expressed concern about the lack of wheels.  As pointed out above, we do publish 32bit windows wheels and I think at least debian is still building the 32bit flavor in their packaging system.

The arm discussion is a bit of a red herring as you would need different wheels from the 32bit x86 one anyway Matplotlib does not upload arm wheels of anykind.

We use @matthew-brett 's wonderful multi-build system to built our mac and linux wheels, however the 32bit linux builds have been broken for a while (I remembered it being related to tests failing due to slight differences in float , but now it looks like we are failing to install dependencies https://travis-ci.org/MacPython/matplotlib-wheels/jobs/650100882).  If those builds get fixed I'm happy to upload the wheels generated.

That said, are we sure that there are 32bit linux users that are installing from pypi on a regular basis?

**Comment by tacaswell** [comment]
> I think it's been true in the past that 32-bit errors and crashes have revealed coding errors that needed fixing for other platforms.

I don't recall any 32bit related errors recently, big-endian vs little endian has been much better at finding actual bugs. 

To expand a bit on my "failing tests" comment above, when we investigated the failures from debian carefully, the 32-bit vs 64-bit differences tended to be things where you had rounding differences due to float precision.  At the end of the day Matplotlib produces RBG(A) values as 4 8bit uints. Every so often you get un-lucky and the rounding error flips a bit in the output which in turn fails our tests (which assert the output images should be bit-identical).

**Comment by alexdesiqueira** [comment]
> I do not think it is fair to say "Matplotlib does not support 32bit", we just have not uploaded 32bit x86 wheels for linux.
> (...) 
>  If those builds get fixed I'm happy to upload the wheels generated.

Thank you @tacaswell, and sorry I didn't ping you here before. Since you're around and cooperating, I guess we don't have a strong reason to drop matplotlib, then... :sweat_smile: 

> (...) but now it looks like we are failing to install dependencies https://travis-ci.org/MacPython/matplotlib-wheels/jobs/650100882).

```
 build/temp.linux-i686-3.7/_openssl.c:546:30: fatal error: openssl/opensslv.h: No such file or directory
```
I don't understand envs on Travis that much (or _at all_), and sorry if this is "raining on wetlands", but it seems this env is missing  [`libssl-dev`](https://launchpad.net/ubuntu/trusty/+package/libssl-dev).

**Comment by hmaarrfk** [comment]
I know arm is different than x86, but I just wanted to make the point that "we fail on 32 bit therefore we should rip it out" is not so good. I'll look into the matplotlib 32 bit wheels at some point. The `--prefer-binary` flag can help find wheels instead of source dists.

**Comment by hmaarrfk** [comment]
some ssl stuff recently changed and broke compatibility. conda has been struggling with this for a while. will try to find some notes.

**Comment by tacaswell** [comment]
@hmaarrfk ah, sorry I misunderstood what you were trying to say about ARM.

**Comment by jni** [comment]
Continuing to build windows 32-bit but not Linux 32-bit is a happy compromise for me! I'll aim to implement that soon.

**Comment by hmaarrfk** [comment]
I'm not trying to create busywork for people, but I had a PR going with the change required to get matplotlib wheels to build by finding wheels instead of src packages for cryptography

https://github.com/MacPython/matplotlib-wheels/pull/9/files

There was a failure due to computational precisions, but that seems like it can be relaxed.

**Comment by hmaarrfk** [comment]
Ok, https://github.com/MacPython/matplotlib-wheels/pull/10 shows that it isn't an isolated failure.

Personally, I think due to our liberal use of cython, it is useful to test on 32bit linux. We find bugs and are generally more careful to note that `int != int64` and `int != int32`.

That said, if a dependency of ours stops shipping wheels, it is a bad sign.

I don't see any problem keeping with the status quo though. What problems, if any does it cause for us? Would you like one build in the build matrix of this repo to be 32 bit linux? We can do that.

**Comment by tacaswell** [comment]
@hmaarrfk Sorry about that PR getting closed / lost.  I did not see it and it got auto-closed when I re-named the branch.  Definitely a case of incompetence rather than malice.

I merged  MacPython/matplotlib-wheels#10 and will backport it to 3.2.x and 3.1.x to see if that un-breaks the builds.

**Comment by hmaarrfk** [comment]
@tacaswell, I totally get it, no malice taken.

We can probably get cryptography to release new wheels too. I'm wondering how much I want to jump down this rabbit hole though.

In my effort to build up the conda-stack for arm, we also decided to drop support for 32 bit arm, even though I know very well it is the more popular platform for embedded systems given the arguments I made above. Narrowing the scope is something we must all do.

**Comment by tacaswell** [comment]
ok, builds are running now, but pytest is still failing and preventing upload, I'll look into fixing working around that tomorrow (suspect the "fix" will be to push a commit disabling running the tests, get the artifact, and then restore the check).

**Comment by hmaarrfk** [comment]
Have you all decided to upload wheels automatically after a successful build?

**Comment by hmaarrfk** [comment]
@tacaswell just as an FYI, we currently pull in the older 32 bit wheels from the wheelhouse. 

If you don't want to support 32 bit matplotlib on Linux, we are happy to keep the status quo going and continue to pull them in from there.

**Comment by tacaswell** [comment]
sorry, did not get to poke at this over the weekend, hopefully tomorrow.

I don't trust uploading directly from travis -> pypi, but they do go to the wheel house.
